core/websock.js

```diff
@@ -173,49 +175,92 @@ export default class Websock {
         this._sQ = new Uint8Array(this._sQbufferSize);
     }
 
+    _getChannelStates(channelType) {
+        if (channelType === "WebSocket") {
+            return {
+                CONNECTING: WebSocket.CONNECTING,
+                OPEN: WebSocket.OPEN,
+                CLOSING: WebSocket.CLOSING,
+                CLOSED: WebSocket.CLOSED
+            };
+        }
+        if (channelType === "RTCDataChannel") {
+            // Constants pulled from RTCDataChannelState enum
+            // https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/readyState#RTCDataChannelState_enum
+            return {
+                CONNECTING: "connecting",
+                OPEN: "open",
+                CLOSING: "closing",
+                CLOSED: "closed"
+            };
+        }
+        throw new Error(`channelType: ${channelType} not recognized`);
+    }
+
     init() {
         this._allocate_buffers();
         this._rQi = 0;
-        this._websocket = null;
+        this._webChannel = null;
     }
 
-    open(uri, protocols) {
+    open({ uri, protocols, webChannel, channelType }) {
```

Comment:
Can't we do "uri or object" here as well?

Reply:
from your earlier comments, if I isolate whether it is a url or existing websocket/datachannel in the constructor of RFB, accepting `open(uriOrWebChannel, protocols)` here would be polluting again and require similar checks to determine if it is a string, or an object, if it is an object, is it a rtcdatachannel/websocket. I can make this change if you want, but this API is internal so we can break it without worry.

Reply:
So far we've been trying to keep this somewhat independent (we are supposed to share it with the websockify repo).

Since it isn't a constructor we could split it to multiple functions for clarity. E.g. an `open()` that takes an URL and a `attach()` that takes an already open object?

Reply:
Sure that works for me.